File: src/lib/scala/Slate.Core/src/main/scala/slate/core/common/mods/Module.scala
/**
  * <slate_header>
  * author: Kishore Reddy
  * url: https://github.com/kishorereddy/scala-slate
  * copyright: 2016 Kishore Reddy
  * license: https://github.com/kishorereddy/scala-slate/blob/master/LICENSE.md
  * desc: a scala micro-framework
  * usage: Please refer to license on github for more info.
  * </slate_header>
  */
package slate.core.common.mods

import slate.common.{Strings, Result}

import scala.collection.mutable.ListBuffer


class Module
{
  var ctx:ModuleContext = null

  /**
    * module info setup on init of derived modules
    */
  var info:ModuleInfo = null


  /**
    * hook to initialize
    */
  def init():Unit = {
  }


  def register():Unit = {

  }


  /**
    * install this module
    */
  def install():Unit = {
    if(info.isDbDependent ){
      info.models.fold(Unit)( models => {
        models.foreach( modelName => {
          ctx.modelService.install(modelName, info.version, "", "")
        })
        Unit
      })
    }
  }


  def script():List[String] = {
    if(info.isDbDependent ){
      info.models.fold(List[String]())( models => {
        models.map(modelName => {
          ctx.modelService.generateSql(modelName, info.version)
          modelName
        })
      })
    }
    else
      List[String]()
  }


  def toItem(): Mod = {
    val item = new Mod()
    item.name          = info.name
    item.desc          = info.desc
    item.version       = info.version
    item.isInstalled   = info.isInstalled
    item.isEnabled     = info.isEnabled
    item.isDbDependent = info.isDbDependent
    item.totalModels   = info.totalModels
    item.source        = info.source
    item.dependencies  = info.dependencies
    item
  }
}

File: src/lib/kotlin/slatekit-integration/src/main/kotlin/slatekit/integration/mods/Module.kt
/**
 * <slate_header>
 * url: www.slatekit.com
 * git: www.github.com/code-helix/slatekit
 * org: www.codehelix.co
 * author: Kishore Reddy
 * copyright: 2016 CodeHelix Solutions Inc.
 * license: refer to website and/or github
 * about: A tool-kit, utility library and server-backend
 * mantra: Simplicity above all else
 * </slate_header>
 */

package slatekit.integration.mods

import slatekit.common.Failure
import slatekit.common.ResultEx
import slatekit.common.Success
import slatekit.common.newline
import slatekit.core.common.AppContext

abstract class Module(
    val appCtx: AppContext,
    val modCtx: ModuleContext
) {
    abstract val info: ModuleInfo

    /**
     * hook to initialize
     */
    open fun init() {}

    open fun register() {}

    /**
     * install this module
     */
    open fun install(): ResultEx<String> {
        return if (info.isDbDependent) {
            info.models?.let { models ->
                val results = models.map { modCtx.setup.install(it, info.version, "", "") }
                val success = results.all { it.success }
                val messages = results.map { it.msg }
                val message = if (success) "" else messages.joinToString(newline)
                if (success) Success(message, msg = "") else Failure(Exception(message), msg = message)
            } ?: Failure(Exception(this.info.name + " has no models"))
        } else {
            Failure(Exception(this.info.name + " is not database dependent"))
        }
    }

    /**
     * install this module
     */
    open fun uninstall(): ResultEx<String> {
        return if (info.isDbDependent) {
            info.models?.let { models ->
                val results = models.map { modCtx.setup.uinstall(it) }
                val success = results.all { it.success }
                val messages = results.map { it.msg }
                val message = if (success) "" else messages.joinToString(newline)
                if (success) Success(message, msg = "") else Failure(Exception(message), msg = message)
            } ?: Failure(Exception(this.info.name + " has no models"))
        } else {
            Failure(Exception(this.info.name + " is not database dependent"))
        }
    }

    open fun seed() {
    }

    /**
     * script out all the sql for all the modules.
     */
    open fun script(): List<String> {
        val result = if (info.isDbDependent) {
            info.models?.let { models ->
                models.map { modelName ->
                    val result = modCtx.setup.generateSql(modelName, info.version)
                    result.msg ?: modelName
                }
            } ?: listOf<String>()
        } else
            listOf<String>()

        return result
    }

    fun toItem(): Mod {
        val item = Mod(
                name = info.name,
                desc = info.desc,
                version = info.version,
                isInstalled = info.isInstalled,
                isEnabled = info.isEnabled,
                isDbDependent = info.isDbDependent,
                totalModels = info.totalModels,
                source = info.source,
                dependencies = info.dependencies
        )
        return item
    }
}
